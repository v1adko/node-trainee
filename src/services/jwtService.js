import jwt from 'jsonwebtoken';
import configJwt from '../config/jwt';

const { tokenSecret: secret, tokenExpiresIn: expiresIn } = configJwt;

const options = { expiresIn };

class JwtService {
  generateJwt = (user) => {
    const payload = {
      _id: user._id,
      role: user.role
    };
    return jwt.sign(payload, secret, options);
  };

  decoder = (token) => {
    const promise = new Promise((resolve, reject) => {
      jwt.verify(token, secret, (error, decoded) => {
        if (error) {
          reject(new Error('Failed to authenticate token.'));
        } else {
          resolve(decoded);
        }
      });
    });
    return promise;
  };
}

export default new JwtService();
