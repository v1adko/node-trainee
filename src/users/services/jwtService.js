const jwt = require('jsonwebtoken');

const { secretTokkenWord: secret } = require('../config/jwt');

const permissionsConst = require('../config/permissions');

class JwtService {
  constructor() {
    this.generateJwt = (user) => {
      const payload = {
        _id: user._id,
        permissions: Object.values(permissionsConst)
      };
      const options = {
        expiresIn: '2h'
      };

      return jwt.sign(payload, secret, options);
    };
  }
}

module.exports = new JwtService();
