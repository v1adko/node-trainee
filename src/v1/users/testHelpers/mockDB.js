import userDao from '../dao';
import userMocks from './userMocks';
import permissions from '../../../constants/permissions';

const { getUsersFields, user: userFields, admin: adminFields } = userMocks;

class MockDB {
  constructor(DAO) {
    this.DAO = DAO;
  }

  createDefaultUsers = async (userCount = 3) => {
    const users = [];
    getUsersFields(userCount).forEach(async (element) => {
      const { username, password, role } = element;
      const user = await this.createUser(username, password, role);
      users.push(user);
    });
  };

  createDefaultUser = async () => {
    const { username, password, role } = userFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createDefaultAdmin = async () => {
    const { username, password, role } = adminFields;
    const user = await this.createUser(username, password, role);
    return user;
  };

  createUser = async (username, password, role = permissions.USER.value) => {
    const createdUser = this.DAO.create(username, password, role);
    return createdUser;
  };

  getAll = async () => {
    const users = await this.DAO.getAll();
    return users;
  };

  cleanDB = async () => {
    await this.DAO.Model.remove();
  };
}

export default new MockDB(userDao);
