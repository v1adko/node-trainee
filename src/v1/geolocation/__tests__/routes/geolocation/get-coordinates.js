import simulate from '../../../../../utils/tests-utils/request-helper';
import jwtService from '../../../../services/jwt-service';
import mockDB from '../../../../../utils/tests-utils/mock-db';
import UserFields from '../../../../../utils/tests-utils/test-user-fields';

jest.setTimeout(10000);

const ROUTE = '/v1/geolocation';

const { username, password } = UserFields;
let userToken = null;

async function clean() {
  await mockDB.cleanDB();
  userToken = null;
}

async function createUser() {
  const user = await mockDB.createUser(username, password);
  userToken = jwtService.generateJwt(user);
}

beforeAll(clean);
afterAll(mockDB.closeConnection);

describe('Test the "/v1/geolocation/:lat/:lon" path', () => {
  beforeEach(createUser);
  afterEach(clean);

  it('should response the GET method', async () => {
    const result = await simulate.get(`${ROUTE}/50/30`, 200, userToken);
    const { address, coordinates } = result[0];
    expect(address).toBe("Unnamed Road, Kyivs'ka oblast, Ukraine");
    expect(coordinates).toEqual({ lat: 49.999137, lon: 30.0019538 });
  });

  it('should response the GET method', async () => {
    const { error, message } = await simulate.get(
      `${ROUTE}/444/444`,
      200,
      userToken
    );
    expect(error).toBe('Error');
    expect(message).toBe('Response status code is 400');
  });

  it('should return error because first argument is not a number', async () => {
    const { message } = await simulate.get(`${ROUTE}/abc/100`, 400, userToken);
    expect(message).toMatchSnapshot();
  });

  it('should return error because second argument is not a number', async () => {
    const { message } = await simulate.get(`${ROUTE}/100/abc`, 400, userToken);
    expect(message).toMatchSnapshot();
  });
});
