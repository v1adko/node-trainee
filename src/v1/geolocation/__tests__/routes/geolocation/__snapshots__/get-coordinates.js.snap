// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the "/v1/geolocation/:lat/:lon" path should return error because first argument is not a number 1`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"lat\\" must be a number",
  "messages": Array [
    "RequestValidationError(400): \\"lat\\" must be a number",
  ],
  "name": "RequestValidationError",
}
`;

exports[`Test the "/v1/geolocation/:lat/:lon" path should return error because second argument is not a number 1`] = `
Object {
  "code": 400,
  "message": "RequestValidationError(400): \\"lon\\" must be a number",
  "messages": Array [
    "RequestValidationError(400): \\"lon\\" must be a number",
  ],
  "name": "RequestValidationError",
}
`;
